coherence/coherence.go

```diff
@@ -0,0 +1,128 @@
+package coherence
+
+/*
+ * Copyright © 2023, Oracle and/or its affiliates.
+ */
+import (
+	"context"
+	coh "github.com/oracle/coherence-go-client/coherence"
+	"time"
+)
+
+const (
+	defaultScopeName = "fiber-session-store"
+	defaultTimeout   = time.Duration(30) * time.Millisecond
+)
+
+// Storage represents an implementation of Coherence storage provider.
+type Storage struct {
+	session    *coh.Session
+	namedCache coh.NamedCache[string, []byte]
+	ctx        context.Context
+}
+
+// Config defines configuration options for Coherence connection.
+type Config struct {
+	// Address to connect to, defaults to "localhost:1408"
+	Address string
+
+	// UseSSL specified if to use SSL or plain text, defaults to false
+	UseSSL bool
+
+	// SessionTimeout is the default session timeout to connect to Coherence, defaults to 30s
+	SessionTimeout time.Duration
+
+	// SessionScope defines a scope allowing for multiple storage sessions
+	SessionScope string
+}
+
+// DefaultConfig defines default options.
+var DefaultConfig = Config{
+	Address:        "localhost:1408",
+	UseSSL:         false,
+	SessionTimeout: time.Duration(30) * time.Millisecond,
+	SessionScope:   defaultScopeName,
+}
+
+// New returns a new [Storage] given a [coherence.Session].
+func New(config ...Config) (*Storage, error) {
+	var (
+		scopeName = defaultScopeName
+		cfg       = DefaultConfig
+	)
+
+	if len(config) == 1 {
+		cfg = config[0]
+	}
+	options := make([]func(session *coh.SessionOptions), 0)
+
+	// apply any config values
+	if cfg.Address != "" {
+		options = append(options, coh.WithAddress(cfg.Address))
+	} else {
+		cfg.Address = DefaultConfig.Address
+	}
+
+	if !cfg.UseSSL {
+		options = append(options, coh.WithPlainText())
+	}
+
+	if cfg.SessionTimeout != defaultTimeout {
+		options = append(options, coh.WithReadyTimeout(cfg.SessionTimeout))
+	}
+
+	if cfg.SessionScope != defaultScopeName {
+		scopeName = cfg.SessionScope
+	}
+
+	// create the Coherence session
+	session, err := coh.NewSession(context.Background(), options...)
+	if err != nil {
+		return nil, err
+	}
+
+	return newCoherenceStorage(session, scopeName)
+}
+
+// newCoherenceStorage returns a new Coherence [Storage].
+func newCoherenceStorage(session *coh.Session, cacheName string) (*Storage, error) {
+	nc, err := coh.GetNamedCache[string, []byte](session, "fiber$"+cacheName)
```

Comment:
Can we make this "fiber$" prefix configurable? In case someone already has a DB with data, that prefix wouldnt let them use it, right?

Reply:
The `Config.ScopeName` is exactly for that reason. (The fiber$ prefix is just for the cache name so that it wouldn't )

If you are for a default coherence Store, the underlying coherence cache will be "fiber$default-store" (which i've changed from fiber-session-store as its not just for sessions).

If this is asked for by difference processes or new vars, they will map to the same underlying coherence cache.

If you want separate stores, you can use
`store1, err := New(Config{ScopeName: "scope1"})`  - which will store data in coherence cache named "fiber$scope1"

`store2, err := New(Config{ScopeName: "scope2"})` - which will store data in coherence cache named "fiber$scope2"

I guess the scopeName is similar to the `Bucket` in couchbase, but we also add a prefix to avoid conflict with other coherence caches.

Hope that makes sense.